feat: Add support for new `BATCH` message type in taps and targets (#904)
meltano/sdk: docs/batch.md

> # PREVIEW - Batch Messages (A.K.A. Fast Sync)

Reviewer: ```suggestion
# Batch Messages
```

As a page title, I think it's best to keep this simple. (I appreciate though that the preview warning is top of page just below the title.)

Also, since "Fast Sync" is technically the Pipelinewise spec name, we should avoid the name in technical documentation since we are not technically compatible with that separate spec - even though we are inspired by it.

```{warning}

> The `BATCH` message functionality is currently in preview and is subject to change.

Reviewer: ```suggestion
The `BATCH` message functionality is currently in preview and is subject to change.
You can open a [GitHub issue](https://github.com/meltano/sdk/issues) or [join the discussion](https://github.com/meltano/sdk/discussions/963) on GitHub to provide feedback during the preview period.
```

> ```{warning}
> The `BATCH` message functionality is currently in preview and is subject to change.
> ```

Reviewer: ```suggestion
```{warning}
The `BATCH` message functionality is currently in preview and is subject to change.
You can open a [GitHub issue](https://github.com/meltano/sdk/issues) or [join the discussion](https://github.com/meltano/sdk/discussions/963) on GitHub to provide feedback during the preview period.
```
```

- when the tap outputs records at a much higher rate than the target can consume them, creating backpressure
- when the source system can directly export data in bulk (e.g. a database dump)

Currently only a local filesystem is supported, but other filesystems like AWS S3, FTP, etc. could be supported in the future.

## The `BATCH` Message

```json
{
  "type": "BATCH",
  "stream": "users",
  "encoding": {
    "format": "jsonl",
    "compression": "gzip"
  },
  "manifest": [
    "path/to/batch/file/1",
    "path/to/batch/file/2"
  ]
}
```

### `encoding`

The `encoding` field is used to specify the format and compression of the batch files. Currently only `jsonl` and `gzip` are supported, respectively.

### `manifest`

The `manifest` field is used to specify the paths to the batch files. The paths are relative to the `root` directory specified in the [`batch_config`](#batch-configuration) storage configuration.

## Batch configuration

The batch configuration is used to specify the root directory for the batch files, and the maximum number of records per batch file.

> ```json
> {
>   "encoding": {
>     "format": "jsonl",
>     "compression": "gzip",
>   },
>   "storage": {
>     "root": "file://tests/core/resources",
>     "prefix": "test-batch-",
>   }
> }

Reviewer: Should this be under a `batch_config` parent item?

```suggestion
```js
{
  // ...
  "batch_config": {
    "encoding": {
      "format": "jsonl",
      "compression": "gzip",
    },
    "storage": {
      "root": "file://tests/core/resources",
      "prefix": "test-batch-",
    }
  }
}
```

However, the Singer specification can be extended to support additional types of messages. For example, the [`ACTIVATE_VERSION`](https://sdk.meltano.com/en/latest/capabilities.html#singer_sdk.helpers.capabilities.PluginCapabilities.ACTIVATE_VERSION) message is used to manage hard deletes in the target.

This library's implementation of the `BATCH` message is used to send records in bulk from the tap to the target, using an intermediate filesystem to store _batch_ files. This is useful, for example

- when the tap outputs records at a much higher rate than the target can consume them, creating backpressure
- when the source system can directly export data in bulk (e.g. a database dump)

Currently only a local filesystem is supported, but other filesystems like AWS S3, FTP, etc. could be supported in the future.

## The `BATCH` Message

```json
{
  "type": "BATCH",
  "stream": "users",
  "encoding": {
    "format": "jsonl",
    "compression": "gzip"
  },

>   "manifest": [
>     "path/to/batch/file/1",
>     "path/to/batch/file/2"
>   ]

Reviewer: I think this edit can be helpful, relevant to recent updates to pass and accept absolute paths:

```suggestion
  "manifest": [
    "file://path/to/batch/file/1",
    "file://path/to/batch/file/2"
  ]
```

}
```

### `encoding`

The `encoding` field is used to specify the format and compression of the batch files. Currently only `jsonl` and `gzip` are supported, respectively.

### `manifest`

The `manifest` field is used to specify the paths to the batch files. The paths are relative to the `root` directory specified in the [`batch_config`](#batch-configuration) storage configuration.

> ## Batch configuration
> 
> The batch configuration is used to specify the root directory for the batch files, and the maximum number of records per batch file.

Reviewer: ```suggestion
## Batch configuration

When local storage is used, targets do no require special configuration to process `BATCH` messages.

Taps may be configured to specify a root storage `root` directory, file path `prefix`, and `encoding` for batch files using a configuration like the below:

```

In `config.json`:

```js
{
  // ...
  "batch_config": {
    "encoding": {
      "format": "jsonl",
      "compression": "gzip",
    },
    "storage": {
      "root": "file://tests/core/resources",
      "prefix": "test-batch-",
    }
  }
}
```

## Custom batch file creation and processing

### Tap side

> The tap can customize the batch file creation by implementing the [`get_batches`](singer_sdk.Stream.get_batches). This method should return a _tuple_ of an encoding and a list of batch files:

Reviewer: ```suggestion
Taps can optionally customize the batch file creation by implementing the [`get_batches`](singer_sdk.Stream.get_batches). This method should return a _tuple_ of an encoding and a list of batch files:
```

```python
class MyStream(Stream):
    def get_batches(self, records):
        return (
            ParquetEncoding(compression="snappy"),
            [
                "s3://my-bucket/my-batch-file-1.parquet",
                "s3://my-bucket/my-batch-file-2.parquet",
            ]
        )
```

### Target side

> The target can customize the batch file processing by implementing the [`process_batch_files`](singer_sdk.Sink.process_batch_files).

Reviewer: ```suggestion
Targets can optionally customize the batch file processing by implementing the [`process_batch_files`](singer_sdk.Sink.process_batch_files).
```
